gdal/frmts/rmf/rmfjpeg.cpp

```diff
@@ -0,0 +1,183 @@
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifdef HAVE_LIBJPEG
+#include <algorithm>
+#include "cpl_conv.h"
+
+#include "rmfdataset.h"
+#include <setjmp.h>
+CPL_C_START
+#include <jpeglib.h>
+CPL_C_END
+
+
+static void RMFJPEGMessage(j_common_ptr poInfo, int iMsgLevel)
+{
+    if(iMsgLevel > 0)
+    {
+        return;
+    }
+
+    poInfo->err->num_warnings++;
+
+    if(poInfo->err->num_warnings > 10)
+    {
+        return;
+    }
+
+    char aszBuffer[JMSG_LENGTH_MAX] = {0};
+
+    poInfo->err->format_message(poInfo, aszBuffer);
+    CPLError(CE_Failure, CPLE_AppDefined, "RMF JPEG:%s", aszBuffer);
+}
+
+static void RMFJPEGError(j_common_ptr poInfo)
+{
+    jmp_buf* poJmpBuf = reinterpret_cast<jmp_buf*>(poInfo->client_data);
+    char     aszBuffer[JMSG_LENGTH_MAX] = {0};
+
+    poInfo->err->format_message(poInfo, aszBuffer);
+    CPLError(CE_Failure, CPLE_AppDefined, "RMF JPEG: %s", aszBuffer);
+    longjmp(*poJmpBuf, 1);
+}
+
+static void RMFJPEGNoop(j_decompress_ptr)
+{
+}
+
+/************************************************************************/
+/*                          JPEGDecompress()                            */
+/************************************************************************/
+
+int RMFDataset::JPEGDecompress(const GByte* pabyIn, GUInt32 nSizeIn,
+                               GByte* pabyOut, GUInt32 nSizeOut,
+                               GUInt32 nRawXSize, GUInt32 nRawYSize)
+{
+    if(pabyIn == nullptr ||
+       pabyOut == nullptr ||
+       nSizeOut < nSizeIn ||
+       nSizeIn < 2)
+       return 0;
+
+    jpeg_decompress_struct  oJpegInfo;
+    jpeg_source_mgr         oSrc;
+    jpeg_error_mgr          oJpegErr;
+    jmp_buf                 oJmpBuf;
+
+    oJpegInfo.err = jpeg_std_error(&oJpegErr);
+    oJpegErr.error_exit = RMFJPEGError;
+    oJpegErr.emit_message = RMFJPEGMessage;
+    oJpegInfo.client_data = reinterpret_cast<void*>(&oJmpBuf);
+
+    memset(&oSrc, 0, sizeof(jpeg_source_mgr));
+
+    oSrc.next_input_byte = (JOCTET *)pabyIn;
+    oSrc.bytes_in_buffer = (size_t)nSizeIn;
+    oSrc.term_source = RMFJPEGNoop;
+    oSrc.init_source = RMFJPEGNoop;
+
+    jpeg_create_decompress(&oJpegInfo);
+
+    memset(&oJmpBuf, 0, sizeof(jmp_buf));
+
+    if(setjmp(oJmpBuf))
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "RMF JPEG: Error decompress JPEG tile");
+        jpeg_destroy_decompress(&oJpegInfo);
+        return 0;
+    }
+
+    oJpegInfo.src = &oSrc;
+    jpeg_read_header(&oJpegInfo, TRUE);
+
+    oJpegInfo.dct_method = JDCT_FLOAT;
+    oJpegInfo.out_color_space = JCS_RGB;
+
+    jpeg_start_decompress(&oJpegInfo);
+
+    JDIMENSION  nImageHeight = std::min(oJpegInfo.image_height,
```

Comment:
shouldn't there be a check that the number of components is 3 ?

Reply:
It is not possible to create files with a number of bands other than 3.
I'm check it at this [point](https://github.com/OSGeo/gdal/pull/691/files/3a41a23f95fa3a74ce4b80baac20a230f3e52573#diff-a59133a9033503c9219f44d776451b8fR1747).
But it is better to check it here for a future.

Reply:
My comment was to deal with hostile files where the JPEG codestream wouldn't match other expectations of the code

Reply:
Yes, it is important.